### ISSUE TITLE
When Many-to-Many Through Associations release? 

### ISSUE THREAD

**Issue by cuongcua90** [balderdashy/waterline#705]
The docs said it is coming soon. while I wait it is released is the alternative way to attack addition attributes?

**Comment by kirkaleze** [comment]
@devinivy  what statement are you using to populate your join model once humans and dogs are populated? Also, is declaring the variables necessary?

**Comment by devinivy** [comment]
@kirkaleze declaring the variables is merely for illustrative purposes.  The populate query would then look quite normal:
`DogCollection.find().populate('humanFriends').exec(function(err, dogs) {});`

But you should also be able to query the junction table:
`FriendshipsCollection.find().exec(function(err, friendships) {});`

**Comment by dmarcelino** [comment]
It would also be good to be able to store data in the join table (association) while saving. Quoting @elennaro (#957):
> title: extra fields on many to many relationships join table 
> It's a widely used approach when we save extra data with relation.
> Like quantity for sold products in order-product relation.
> Like expiration on user-post-ban. etc. etc.

**Comment by devinivy** [comment]
Yes, indeedy!  Info on the join table should also be included appear when populating records during a `find`.

**Comment by elennaro** [comment]
Sorry for a noob question, but is there any manual way to overwrite find for humans somehow to get both Dogs and identity field from friendship?

**Comment by dmarcelino** [comment]
One of the things I'm really missing is the ability of pulling the join table record for each `through` association. I was thinking about a possible API for this and I came up with:
``` javascript
// joinRecord or some other keyword to enable this behaviour
Model.findOne(100).populate('assocAttribute', { joinRecord: true })
```
Would return something like:
``` javascript
{
  id: 100,
  assocAttribute: [
    { id: 2, name: 'blah', createdAt: '04/05/2015', _joinRecord: { id: 10, createdAt: '05/05/2015', joinAttribute: 'something' } }
  ]
}
```

What do you think about this API? Does anybody knows how other ORMs handle this?

**Comment by chrisns** [comment]
@dmarcelino that API looks similar to the suggested approach in Rails with the  `proxy_association` ActiveRecord Association Extension (from my very limited Rails understanding).

It would be amazing as part of the extension to the `.populate()` api to be able to have criteria on the join, e.g. `createdAt >  04/04/2015`
maybe:

```
# To get all
Model.findOne(100).populate('assocAttribute', { joinRecord: true })
```
and
```
# to get joins made since a date *criteria syntax is probably wrong
Model.findOne(100).populate('assocAttribute', { joinRecord: {where: {createdAt: {'>=': '04/04/2015'}}}})
```
and further to optimize for large join tables, one could also populate only the fields you were interested in maybe:
```
# to get only id from join record 
Model.findOne(100).populate('assocAttribute', { joinRecord: { select: ['id']}})
# to return 
{
  id: 100,
  assocAttribute: [
    { id: 2, name: 'blah', createdAt: '04/05/2015', _joinRecord: { id: 10} }
  ]
}
```

That all said, I'd be extremely happy to get the implementation you suggested for now :smile:  :+1:

**Comment by jettwein** [comment]
Anyone have a working example of this that they'd care to share? I can't seem to get this working properly. I'm sure I'm missing something quite simple.

**Comment by Benew** [comment]
+1 Need this feature desperately. Did anyone manually implement it with success?

**Comment by ksylvan** [comment]
:+1:  Please.

**Comment by nicco** [comment]
I'm using 0.10.26 and all seems to work fine. I have a question thou how can you populate someOtherAttr in Friendship? is there a way to do it from one of the 2 tables part of the relationship or I do need to call explicitly Friendship?

**Comment by elennaro** [comment]
Hello.
Any progress here?
Any help needed?

**Comment by sailsbot** [comment]
Thanks for posting, @cuongcua90.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Reopening, as this is in a feature request in an active milestone.  We can close one it's PRed into the roadmap.

**Comment by mukk85** [comment]
@devinivy Any ETA on when this will be available? Any patches we can integrate today?

**Comment by devinivy** [comment]
@mukk85 this is essentially usable– the configuration for it is just ugly and it's considered experimental.  @atiertant might have a patch that you can integrate?

**Comment by atiertant** [comment]
@devinivy got some work on it,got something working a little,need more test...
there is an other problem:

      {
        identity: 'Taxi',
        connection: 'foo',
        tableName: 'taxi_table',
        attributes: {
          "taxiId": {
            "type": "integer",
            "primaryKey": true
          },
          "taxiMatricule": {
            "type": "string"
          },
          "drivers": {
            "collection": "Driver",
            "through": "ride",
            "via": "taxi"
          }
        }
      }

      {
        identity: 'Ride',
        connection: 'foo',
        tableName: 'ride_table',
        attributes: {
          "rideId": {
            "type": "integer",
            "primaryKey": true
          },
          "taxi": {
            "model": "Taxi"
          },
          "new_driver": {
            "model": "Driver"
          },
          "last_driver": {
            "model": "Driver"
          },
        }
      }

like this this not possible to define if relation on taxi should use  `new_driver` or `last_driver` on ride  ...
maybe spec could had `forward` key to specify when this ambiguous like this : 

          "drivers": {
            "collection": "Driver",
            "through": "ride",
            "via": "taxi",
            "forward": "new_driver"
          }

**Comment by atiertant** [comment]
@particlebanana merges patch to correct many bugs on this assocations type but not this one.
@particlebanana what do you think about adding optional `forward` key ?

**Comment by sailsbot** [comment]
Thanks for posting, @cuongcua90.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
Re-opening until this is added to the roadmap

**Comment by atiertant** [comment]
@particlebanana the real problem is we don't know what must be done... update doc? correct a bug ? add some tests,...just add a line in roadmap to say "something like this should work later"

**Comment by particlebanana** [comment]
Ok this is on the Roadmap. Waterline is taking the contribution guide from Sails. So basically what needs to happen is the following:

> * First, write out a high-level summary of the feature you are proposing as a concise description (3-5 sentences) focused around a convincing real-world use case where the app you are building or maintaining for your job, your clients, your company, your non-profit work, or your independent hobby project would be made easier by this feature or change.
* Next, describe in clear prose with relevant links to code files exactly why it would be difficult or impossible to implement the feature without changing Waterline core. If this is not the case, and this feature could be implemented as an adapter, then please reconsider writing your proposal (it is unlikely the core team will be able to accept it).
* Finally, if you have time, take a first pass at proposing a spec for this feature (its configuration, usage, and how it would be implemented). If you do not have time to write out a first draft of a thorough specification, please make that point in your feature request, and clarify that it would be up to other contributors with the same or a similar use case to finish this proposal.

This has a somewhat defined spec with the `forward` key and we are transitioning over to the new contribution guide for features so I just added it.

So TLDR: We have a loose spec of how to do M:M through with extra columns and it has been added to the roadmap. It could be scheduled for `0.11` or beyond depending on time constraints.
